Dynamic Dashboard: Reviews Card Reload Functionality Part 2

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #12742
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds more functionalities to the reviews card:
1. Filtering reviews by All / Approved / Hold (includes updating `ProductReviewAction.synchronizeProductReviews` to accept a status parameter)
2. Displaying read/unread state of the reviews (shown in the bubble icon color), by adding notification fetching.
3. Tapping individual review in the card to open its details screen.
4. Moves logic for "View all reviews" button navigation to View Controller.

This PR does not include these, they will be added in a separate PR:
1. Proper empty/loading/error states.
2. unit tests
3. Tracking

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Start with a WordPress.com test site with a variety of review statuses (specifically approved and hold).
2. Go to the test site and write a new review,
3. Start app,
5. Check Reviews card (enable first in Dashboard Edit screen if you haven't),
6. Ensure that the new review added on step 2 is shown with blue speech bubble icon,
7. Ensure that the other reviews are shown in blue or gray bubble icon, depending if they have been read or not.
8. Tap a review, ensure this open the review's details,
9. Go back to Dashboard,
10. Test the All / Approved / Hold filtering (please ignore the loading visual for now). Ensure that the shown reviews are correct, and that the label next to "Status" on the Dashboard shows the correct filtering.
11. Test the "View all reviews" button and ensure it still works well.

## Video:

https://github.com/woocommerce/woocommerce-ios/assets/266376/2244ac33-fd1b-4d0c-babd-875bd76a503e

hafizrahman committed:
Functionality to tap individual review to see details.

hafizrahman committed:
Refactor "view all reviews" navigation to also do it via hosting controller.

hafizrahman committed:
Set read state of reviews in card by loading notifications related to it.

This might sound unintuitive but it follows the existing logic in the reviews screen.

hafizrahman committed:
Merge branch 'feat/12742-review-card-functionality'

Conflicts:
	WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift

hafizrahman committed:
Update remote load reviews function to optionally accept filter.

hafizrahman committed:
Add mapping functionality from ReviewsFilter to ProductReviewStatus.

hafizrahman committed:
Update Store/Action to support passing status.

hafizrahman committed:
Add filtering functionality.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman commented on Networking/Networking/Remote/ProductReviewsRemote.swift:
```diff
@@ -26,23 +26,24 @@ public final class ProductReviewsRemote: Remote, ProductReviewsRemoteProtocol {
     ///     - pageNumber: Number of page that should be retrieved.
     ///     - pageSize: Number of Orders to be retrieved per page.
     ///     - products: Limit result set to reviews assigned to specific product IDs.
+    ///     - filterByStatuses: Array of statuses to filter reviews by.
```
This is an array because filtering by multiple statuses is supported. I'm copying FluxC's implementation here:

https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/14fcce73130015f14125e9e13837f84ff4171bb3/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt#L1584

hafizrahman commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift:
```diff
@@ -105,43 +108,52 @@ private extension ReviewsDashboardCard {
-                        .bodyStyle()
-                        .padding(.trailing, Layout.padding)
-                }
+        Button {
```
Big diff in this chunk but it's essentially just:
1. Wrapping the previous View into Button so it's tappable
2. Small logic change on how bubble icon color is chosen (now based on read status instead of pending status)
3. Adding `.multilineTextAlignment` in some places because wrapping them in Button requires doing to avoid text being centered.

itsmeichigo replied:
Some nit-picking: the top and bottom padding of the review item seem to be uneven:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/1c8fa987-62d3-42eb-bb52-62dd7966e8d2" width=320 />

The reason is that you added a padding to the divider that is rendered only for non-last items.

I suggest using padding on the parent items rather than the child items. That would ensure all the padding is consistent and simplify the code, making it easier to parse. Something like this:

```Swift
Button {
    onViewReviewDetail(viewModel)
} label: {
    HStack(alignment: .top, spacing: Layout.padding) {
         Image(...) // No padding
         VStack(...) // No padding for any child items
    }
    .padding(.horizontal, Layout.padding)
    .padding(.vertical, Layout.cardPadding)
}
```

hafizrahman replied:
The purpose for adding individual padding in the child items of the review content `VStack` is because the subject and review text needs padding on the right so the text doesn't stick to the right edge, but the Divider needs to stretch and stick to the right edge:

<img width="421" alt="Screenshot 2024-05-21 at 11 50 39" src="https://github.com/woocommerce/woocommerce-ios/assets/266376/3b607ddc-5204-4a65-888f-19938332b99f">

I can't figure out those padding differences without specifying individual padding for each. I suppose I can group the non-divider child items into their own `VStack` and set consistent padding for them there.

It's also the same story with the Image. It wants its own vertical padding because without it, it sticks a bit too high compared to the text next to it, like this:

<img width="95" alt="Screenshot 2024-05-21 at 11 49 25" src="https://github.com/woocommerce/woocommerce-ios/assets/266376/a4158b20-ffd5-4dbb-8193-8b876ad4daca">


Let me think about it some more.

itsmeichigo replied:
Good points, thanks for the explanations!

> The purpose for adding individual padding in the child items of the review content VStack is because the subject and review text needs padding on the right so the text doesn't stick to the right edge, but the Divider needs to stretch and stick to the right edge:

I didn't realize this. In this case, you should probably group the review content and divider within a VStack and add the bottom padding to it. That way you will not lose the padding when the divider is not displayed. 

> I can't figure out those padding differences without specifying individual padding for each. I suppose I can group the non-divider child items into their own VStack and set consistent padding for them there.

Usually, we use the default padding 16 for horizontal padding. The text might be adjusted a bit due to line break, so it doesn't look strictly 16 from the design.

> It's also the same story with the Image. It wants its own vertical padding because without it, it sticks a bit too high compared to the text next to it, like this:

You're using `.top` alignment for `HStack`. Try `.firstTextBaseline` to see if it helps.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -59,6 +63,24 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
     }()
 
+    /// ResultsController for Notification
+    private lazy var notificationsResultsController: ResultsController<StorageNote> = {
+        let sortDescriptor = NSSortDescriptor(keyPath: \StorageNote.timestamp, ascending: false)
+        return ResultsController<StorageNote>(storageManager: storageManager,
+                                              sectionNameKeyPath: "normalizedAgeAsString",
+                                              matching: notificationsPredicate,
+                                              sortedBy: [sortDescriptor])
+    }()
+
+    private lazy var notificationsPredicate: NSPredicate = {
+        let notDeletedPredicate = NSPredicate(format: "deleteInProgress == NO")
+        let typeReviewPredicate = NSPredicate(format: "subtype == %@", Note.Subkind.storeReview.rawValue)
+
+        return NSCompoundPredicate(andPredicateWithSubpredicates: [typeReviewPredicate,
+                                                                   sitePredicate(),
+                                                                   notDeletedPredicate])
+    }()
+
```
This is simply the same as existing below:

https://github.com/woocommerce/woocommerce-ios/blob/48d44a8560b6c6b8fa27f4f0fae0af83699dfa1b/WooCommerce/Classes/ViewRelated/Reviews/ReviewsDataSource.swift#L43

hafizrahman commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -147,8 +181,14 @@ private extension ReviewsDashboardCardViewModel {
-                // TODO: also fetch notification
-                return ReviewViewModel(review: review, product: product, notification: nil)
+                let notification = notifications.first { notification in
+                    if let notificationReviewID = notification.meta.identifier(forKey: .comment) {
```
Same matching logic as in https://github.com/woocommerce/woocommerce-ios/blob/48d44a8560b6c6b8fa27f4f0fae0af83699dfa1b/WooCommerce/Classes/ViewRelated/Reviews/ReviewsDataSource.swift#L194

hafizrahman committed:
Merge branch 'trunk'

Conflicts:
	WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift

hafizrahman committed:
Update functions post-merge

hafizrahman committed:
Add menu selected state.

hafizrahman committed:
Update notification results controller behavior to be like products results behavior.

hafizrahman committed:
Update logic for showing or hiding all reviews button.

hafizrahman committed:
update existing unit tests

hafizrahman committed:
Fix more unit test error.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift:
```diff
@@ -261,6 +262,27 @@ private extension DashboardViewHostingController {
+            let view = ReviewDetailView(productReview: review.review,
+                                        product: review.product,
+                                        notification: review.notification)
+            let hostingController = UIHostingController(rootView: view)
```
Here we are using UIKit to show the view, so we should use `ReviewDetailsViewController` directly instead of pushing a hosting controller of a wrapper of the view controller. Simplifying the view hierarchy can save us from unexpected bugs and optimize performance.

hafizrahman replied:
Thanks, I learned something new. Added in 09297efb7394d5a6b094aeee94a9de25c501847e

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift:
```diff
@@ -261,6 +262,27 @@ private extension DashboardViewHostingController {
+        rootView.onViewAllReviews = { [weak self] in
+            guard let self else { return }
+            let view = ReviewsView(siteID: viewModel.siteID)
+            let hostingController = UIHostingController(rootView: view)
```
Similar comment as above but for `ReviewsViewController`.

hafizrahman replied:
Added in 09297efb7394d5a6b094aeee94a9de25c501847e

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -48,6 +54,11 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
                                                                            sortedBy: [])
+
+        self.notificationsResultsController = ResultsController<StorageNote>(storageManager: storageManager,
+                                                                           sectionNameKeyPath: "normalizedAgeAsString",
```
Is it necessary?

hafizrahman replied:
This entire ResultsController was copied exactly from `ReviewsDataSource`

https://github.com/woocommerce/woocommerce-ios/blob/0235248c87238e55c674d014c425a05870ada14e/WooCommerce/Classes/ViewRelated/Reviews/ReviewsDataSource.swift#L48

The particular code was initially added almost six years ago in this commit https://github.com/woocommerce/woocommerce-ios/commit/01b0281c0581d8b6c9a889979ecc8037ec15b9ac

I did not investigate if it's deprecated along the way and needs to be removed. If it is, `ReviewsDataSource` will need to be updated too (perhaps we can make a separate issue about it).

itsmeichigo replied:
I think it's necessary to understand what `sectionNameKeyPath` is for. You can look it up for more detailed explanations, but it's essentially used to group the models into sections by a specified key path. This is necessary for the reviews list because we're displaying the list in sections based on the age of the reviews. In our case, we display the reviews list without grouping them, so using `sectionNameKeyPath` is redundant.

hafizrahman replied:
Yup! I did eventually investigate yesterday and removing it seems to remove the issue (mentioned this in our Slack discussion).

itsmeichigo reviewed (commented):
From my testing, the reviews are still marked as unread even though they are read. It seems that the logic to fetch product details and notifications is not working properly:


https://github.com/woocommerce/woocommerce-ios/assets/5533851/249d5248-2e2b-435a-8d51-86bea78ad2cb

Looking at the code, I see a few problems:
- You are not observing any changes to the product results controller and notification results controller. So after querying items from storage and not finding them, you sync products and notifications but the upserted items are not used to update the UI. That's potentially why all the reviews are unread for me.
- When I filter pending reviews, open one item and approve it, I expect it to be removed from the card without pulling-to-refresh. This would work if you observe the view controllers' changes.

The tangling logic between fetching remote and querying from storage doesn't work well here, unless you have clear steps to minimize the network requests, I suggest following the logic in `ReviewsViewModel` for now. Which means:
- Following [`synchronizeReviews`](https://github.com/woocommerce/woocommerce-ios/blob/f6f0b220d4dfac7a0a2bdb727dabc1bc682eeb4c/WooCommerce/Classes/ViewRelated/Reviews/ReviewsViewModel.swift#L124) method to trigger remote requests in one place. Use a task group to sync both reviews and notifications at the same time.
- Rely solely on results controllers to update the UI.

hafizrahman commented:
Thanks for the detailed review @itsmeichigo . The suggestion to improve the logic is great. To address this item specifically:

> You are not observing any changes to the product results controller and notification results controller. So after querying items from storage and not finding them, you sync products and notifications but the upserted items are not used to update the UI. That's potentially why all the reviews are unread for me.

The reason product results are not observed is [from my implementation of a suggestion on this previous PR.](https://github.com/woocommerce/woocommerce-ios/pull/12752/commits/4ad2e89c4dcf4bd4c51f0ba8662446e96aad4ccd#r1604518766).

In there previously, the product results controller is the usual `lazy var` and had a `configure` function for `onDidResetContent` and `onDidChangeContent`. The product results controller require a specific parameter (product IDs) to be available for its predicate, yet configuring it during init then makes the predicate wrong since product ID isn't available yet at the time.

So instead of observing, the code pulls the results controller directly here:

https://github.com/woocommerce/woocommerce-ios/pull/12769/files#diff-72687964fa287c936bf60846c303ba6b97085af2493b322103838b0d334c7961R193-R194

```
        data = reviews
            .map { review in
                let product = reviewProducts.first { $0.productID == review.productID }
                let notification = notifications.first { notification in
                    if let notificationReviewID = notification.meta.identifier(forKey: .comment) {
                        return notificationReviewID == review.reviewID
                    }
                    return false
                }

                return ReviewViewModel(review: review, product: product, notification: notification)
            }
```

`reviewsProducts` and `notifications` are computed var that do `.fetchedObjects` for each result.

> When I filter pending reviews, open one item and approve it, I expect it to be removed from the card without pulling-to-refresh. This would work if you observe the view controllers' changes.

Yup, this I haven't added in this PR, and I understand that adding observation will fix it. 

> The tangling logic between fetching remote and querying from storage doesn't work well here, unless you have clear steps to minimize the network requests, I suggest following the logic in ReviewsViewModel for now.

I have introduced more logic to also show partial result in the next (draft) PR #12787 . I'll add the refactor suggestion here and include that partial result as well so it can be reviewed in one go.

itsmeichigo commented:
> reviewsProducts and notifications are computed var that do .fetchedObjects for each result.

If I'm not mistaken, `fetchedObjects` are not updated unless you call `performFetch` again after updating the query's predicate. Could you confirm that please?

hafizrahman commented:
> If I'm not mistaken, fetchedObjects are not updated unless you call performFetch again after updating the query's predicate. Could you confirm that please?

Yup, this is correct. In the logic for `updateReviewsResults()`, the code is like this:

```
            updateProductsResultsControllerPredicate(with: productIDs)
            do {
                try await fetchProducts(for: productIDs)
```

and `fetchProducts(for:)` already has `try productsResultsController.performFetch()` in it, so it seems the order is already correct (update predicate first, then call `performFetch` again).

itsmeichigo commented:
> and fetchProducts(for:) already has try productsResultsController.performFetch() in it, so it seems the order is already correct (update predicate first, then call performFetch again).

What I'm seeing is that even though you call `performFetch` again, you call it before `loadReviewProducts` finishes, so any new products from the remote request would not be present in `fetchedObjects`.

hafizrahman committed:
First attempt to refactor the reviews sync process.

hafizrahman committed:
rename functions

hafizrahman committed:
Merge branch 'trunk'

hafizrahman committed:
Improve navigation by using view controllers.

hafizrahman committed:
Remove unneeded sectionNameKeyPath.

hafizrahman committed:
Prefer to show bubble as gray if there' s no notification too (e.g: on initial load).

hafizrahman committed:
Simplify error handling for various calls.

hafizrahman commented:
@itsmeichigo Thanks for your feedbacks! I did the refactor and added some fixes and things seem to work well on my side. But just as before, not 100% sure if it will work consistently on your side. Please re-test both from fresh login state, and also from existing login state (e.g: switch from store A to B then back to A, and check the card in store A).

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift:
```diff
@@ -261,6 +262,25 @@ private extension DashboardViewHostingController {
+            let viewController = ReviewDetailsViewController(productReview: review.review,
+                                                             product: review.product,
+                                                             notification: review.notification)
+            self.show(viewController, sender: self)
```
Nit:

```suggestion
            guard let self else { return }
            let viewController = ReviewDetailsViewController(productReview: review.review,
                                                             product: review.product,
                                                             notification: review.notification)
            show(viewController, sender: self)
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift:
```diff
@@ -261,6 +262,25 @@ private extension DashboardViewHostingController {
+        rootView.onViewAllReviews = { [weak self] in
+            guard let self = self else { return }
+            let viewController = ReviewsViewController(siteID: viewModel.siteID)
+            self.show(viewController, sender: self)
```
Nit:

```suggestion
            guard let self else { return }
            let viewController = ReviewsViewController(siteID: viewModel.siteID)
            show(viewController, sender: self)
```

hafizrahman committed:
Refactor view to reduce individual padding uses.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -86,105 +117,162 @@ private extension ReviewsDashboardCardViewModel {
+    /// Populates data from the local storage for the three main data (reviews, products, notifications).
+    ///
+    func updateData() {
+        let localReviews = reviews.prefix(Constants.numberOfItems)
```
You might want to filter the reviews by the selected status before getting the first 3 items to avoid showing incorrect items after navigating to the reviews list.

hafizrahman replied:
This is a great catch. updated https://github.com/woocommerce/woocommerce-ios/pull/12769/commits/6aa50c86c0a97e9e7b5fe397118e08559aeaa905

hafizrahman replied:
Added additional filtering rule on https://github.com/woocommerce/woocommerce-ios/pull/12769/commits/5fdfd4b6d7ae2cdf26e3ef495e64ae8d0bbe758c (otherwise it might not show things if filter is all).

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -86,105 +117,162 @@ private extension ReviewsDashboardCardViewModel {
+    func synchronizeReviews(filter: ReviewsFilter) async throws {
+        try await synchronizeProductReviews(filter: filter)
+
+        let productIDs = reviews.prefix(Constants.numberOfItems).map { $0.productID }
```
Mixing storage models when working with network requests can cause flakiness. Here you should work with the items from `synchronizeProductReviews` instead. 

I see that in `synchronizeProductReviews` you ignore the items to use storage as a single source of truth - this only works if you don't have other network requests depending on this request.

hafizrahman replied:
> I see that in synchronizeProductReviews you ignore the items to use storage as a single source of truth - this only works if you don't have other network requests depending on this request.

Thanks, learned something new about this pattern. Using the first network result for the second request makes sense. Added in https://github.com/woocommerce/woocommerce-ios/pull/12769/commits/3d301e3d2dfeee06c66778271bd5da70168af24e

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -86,105 +117,162 @@ private extension ReviewsDashboardCardViewModel {
+    func updateData() {
+        let localReviews = reviews.prefix(Constants.numberOfItems)
+
+        // We can show partial review content as long as there are reviews found in storage.
+        if localReviews.isEmpty == false {
+            showLoadingAnimation = false
+        }
```
We should follow the "separation of concern" principle here. It's better to not update the UI while working on data. Instead, you can observe changes in `data` and update `showLoadingAnimation` to false when it's not empty.

hafizrahman replied:
TIL for me, updated in https://github.com/woocommerce/woocommerce-ios/pull/12769/commits/4ba6ea01b764696f6722e197be376018d7e6e4b1

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -86,105 +117,162 @@ private extension ReviewsDashboardCardViewModel {
-                ServiceLocator.crashLogging.logError(error)
-            }
-        }
+            // As long as the app is able to fetch one review once, then the button should always appear.
+            // Later on, when filtering by hold or pending status, the reviews result might become empty.
+            // In that case, the app should keep showing the button to allow the user to see all non-filtered reviews.
+            shouldShowAllReviewsButton = true
```
Similar comment as above about separation of concern. For simplicity, I'd suggest always keeping the button available.

hafizrahman replied:
Ah yup, I just checked that other cards keep the button available even when nothing is available. Updated https://github.com/woocommerce/woocommerce-ios/pull/12769/commits/b61a44ae35b91ea7d9014effc6589aa511b3c495

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -86,105 +117,162 @@ private extension ReviewsDashboardCardViewModel {
-        let allProductsAvailable = productIDs.allSatisfy { productID in
-            reviewProducts.contains { $0.productID == productID }
-        }
+    func updateProducts(productIDs: [Int64]) async throws {
```
This is not used anymore, do we need to remove it?

itsmeichigo replied:
Reading the code again, I think you want to remove `retrieveProducts` from this call and trigger this method from `updateData()`?

hafizrahman replied:
Thanks! Updated this in https://github.com/woocommerce/woocommerce-ios/pull/12769/commits/c3684aeefbc28e9be37bfd67644257ba6879a737

It's supposed to be for calling performFetch() with the latest productID data. I missed re-adding it while updating the error handling.

hafizrahman committed:
self refactor

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
self refactor 2

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman committed:
Opt to always show all reviews button even if there's no results.

hafizrahman committed:
Use the result of first network request (reviews) for the next request (products).

hafizrahman committed:
Filter storage result with current filter to avoid displaying wrong result after loading reviews list.

hafizrahman committed:
Only filter local data if current filter is not all.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/204">18.8</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
update productsresults predicate with productID data and do fetch.

hafizrahman committed:
Observe content of data to decide whether to show loading animation

hafizrahman committed:
Merge branch 'trunk'

hafizrahman committed:
Add error logging.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift:
```diff
@@ -23,33 +27,30 @@ struct ReviewsDashboardCard: View {
-                    .redacted(reason: viewModel.syncingData ? [.placeholder] : [])
-                    .shimmering(active: viewModel.syncingData)
-            }
+            viewAllReviewsButton
+                .padding(.horizontal, Layout.padding)
+                .redacted(reason: viewModel.showLoadingAnimation ? [.placeholder] : [])
+                .shimmering(active: viewModel.showLoadingAnimation)
```
We should maybe render the view all button only if there is no syncing error.

hafizrahman replied:
You're right. This (and the rest of error handling) is already done in the next states-related PR: 

https://github.com/woocommerce/woocommerce-ios/pull/12806

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift:
```diff
@@ -105,44 +108,52 @@ private extension ReviewsDashboardCard {
+                        if viewModel.review.rating > 0 {
+                            HStack(spacing: Layout.starRatingSpacing) {
+                                ForEach(0..<viewModel.review.rating, id: \.self) { _ in
+                                    Image(systemName: "star.fill")
+                                        .resizable()
+                                        .frame(width: Constants.starSize * scale, height: Constants.starSize * scale)
```
Something I missed in the last reviews: the stars are purple while it should be black on the design.

hafizrahman replied:
Good eye! This was already black in the design PR https://github.com/woocommerce/woocommerce-ios/pull/12731 🤔 

Color added in https://github.com/woocommerce/woocommerce-ios/pull/12769/commits/4db65d5d9c13e28fb59aa145c79e2637a66b45aa

itsmeichigo replied:
Thanks! This probably happened because the view is now inside a `Button` - the labels are accent color by default.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -19,11 +19,22 @@ final class ReviewsDashboardCardViewModel: ObservableObject {
-    @Published private(set) var data: [ReviewViewModel] = []
+    @Published private(set) var data: [ReviewViewModel] = [] {
+        didSet {
+            showLoadingAnimation = data.isEmpty
```
It's not safe to assume that we should show animation when there is no data because we don't want to display the loading animation in the empty state. Even though this doesn't happen from our testing because you also set `showLoadingAnimation = false` after the network requests finish. 

Mutating a variable in multiple places makes it hard to follow the logic. My suggestion:
- Restore `syncingData` and update its state in `reloadData` as before.
- Observe `syncingData` and `data` using combine:

```Swift
$syncingData.combineLatest($data)
    .map { syncing, data -> Bool in 
        syncing && data.isEmpty
    }
    .assignTo(&$showLoadingAnimation)
```

itsmeichigo replied:
Another way is keeping `showLoadingAnimation` a computed variable and `syncingData` as a private `@Published`. In that case, `showLoadingAnimation` will be updated when either `syncingData` or `data` is updated, which triggers the SwiftUI view to re-render.

hafizrahman replied:
Tthe combine example works very nicely.

In my test, it also creates a side effect of not showing loading animation when switching status, if there's already local data for reviews with hold/approved status. This is expected since we want to immediately show local data, but in practice the switch requires a second or so to complete, so without loading animation the waiting looks a bit jarring.

As a solution, I'm adding a new variable `switchingStatus` and make the status text shimmer during switching action, video below:

https://github.com/woocommerce/woocommerce-ios/assets/266376/4c77ca51-598f-4771-b615-5a48e701426a

At first I tested just making the filter text ("All"/"Hold"/"Approved") be shimmering, but it doesn't look obvious enough, especially for the short "All" text. So I make all text ("Status: All") to shimmer instead.

hafizrahman replied:
Added combine and status switch animation on https://github.com/woocommerce/woocommerce-ios/pull/12769/commits/4dd707bc6c8afb34f0170d9728718bab94464309

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -86,105 +121,166 @@ private extension ReviewsDashboardCardViewModel {
+    func synchronizeReviews(filter: ReviewsFilter) async throws {
+        let fetchedReviews = try await synchronizeProductReviews(filter: filter)
+
+        let productIDs = fetchedReviews.prefix(Constants.numberOfItems).map { $0.productID }
```
Nit: using `prefix` is not necessary because you already defined the number of items in `synchronizeProductReviews` request.

hafizrahman replied:
You're right as this from a network request. Updated in https://github.com/woocommerce/woocommerce-ios/pull/12769/commits/595c389cf45c1e207c5912f221b5173b79b6a692

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModel.swift:
```diff
@@ -86,105 +121,166 @@ private extension ReviewsDashboardCardViewModel {
+                                                                          pageSize: Constants.numberOfItems,
+                                                                          status: currentFilter.productReviewStatus
+                                                                         ) { result in
+                switch result {
+                case .success(let reviews):
+                    continuation.resume(returning: reviews)
+                case .failure(let error):
+                    continuation.resume(throwing: error)
+                }
```
Nit: this can be simplified with `continuation.resume(with: result)`.

hafizrahman replied:
Nice! Updated in https://github.com/woocommerce/woocommerce-ios/pull/12769/commits/595c389cf45c1e207c5912f221b5173b79b6a692

itsmeichigo reviewed (commented):
Thanks for the updates! The reloading logic seems correct now - I left a few more nit-picking comments but other than that, the PR should be good to go!

hafizrahman committed:
Code simplifications and clean ups.

hafizrahman committed:
Simplify `showLoadingAnimation` handling using combine and introduce animation for switching.

See conversation here for more details and reasoning for switching animation:

https://github.com/woocommerce/woocommerce-ios/pull/12769#discussion_r1610130928

hafizrahman committed:
Update star color matching design.

itsmeichigo reviewed (approved):
Looks good, thanks for the patience!

I'm still unsure if it's beneficial for users to see the reviews before the product details are available:


https://github.com/woocommerce/woocommerce-ios/assets/5533851/aba09e12-7f88-4edc-a319-b3145e9d61e4




 I'll leave the decision to you and pre-approve this PR for now.

hafizrahman commented:
> I'm still unsure if it's beneficial for users to see the reviews before the product details are available:

Thanks for your careful reviews! Let's keep things like this for now and monitor user feedbacks. Changing it to be in load state until products are loaded should not be too time consuming if we need to refactor.

hafizrahman committed:
Merge branch 'trunk'

Conflicts:
	WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift
	WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift

itsmeichigo commented:
@hafizrahman - seems like a unit test for the dashboard cards need to be updated after this PR.

hafizrahman committed:
Update unit tests to include new card (.lastOrders)
